tensorflow_graphics/nn/layer/pointnet.py

```diff
@@ -0,0 +1,252 @@
+#Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Implementation of the PointNet networks from:
+
+@inproceedings{qi2017pointnet,
+  title={Pointnet: Deep learning on point sets
+         for3d classification and segmentation},
+  author={Qi, Charles R and Su, Hao and Mo, Kaichun and Guibas, Leonidas J},
+  booktitle={Proceedings of the IEEE conference on computer vision and pattern
+             recognition},
+  pages={652--660},
+  year={2017}}
+
+NOTE: scheduling of batchnorm momentum currently not available in keras. However
+experimentally, using the batch norm from Keras resulted in better test accuracy
+(+1.5%) than the author's [custom batch norm
+version](https://github.com/charlesq34/pointnet/blob/master/utils/tf_util.py)
+even when coupled with batchnorm momentum decay. Further, note the author's
+version is actually performing a "global normalization", as mentioned in the
+[tf.nn.moments documentation]
+(https://www.tensorflow.org/api_docs/python/tf/nn/moments).
+
+This shorthand notation is used throughout this module:
+  `B`: Number of elements in a batch.
+  `N`: The number of points in the point set.
+  `D`: Number of dimensions (e.g. 2 for 2D, 3 for 3D).
+  `C`: The number of feature channels.
+"""
+
+import tensorflow as tf
+from tensorflow.keras.layers import Layer
+from tensorflow.keras.layers import Dense
+from tensorflow.keras.layers import Conv2D
+from tensorflow.keras.layers import Dropout
+from tensorflow.keras.layers import BatchNormalization
+
+
+class PointNetConv2Layer(Layer):
+  """The 2D convolution layer used by the feature encoder in PointNet."""
+
+  def __init__(self, channels, momentum):
+    """Constructs a Conv2 layer.
+
+    Note:
+      Differently from the standard Keras Conv2 layer, the order of ops is:
+      1. fully connected layer
+      2. batch normalization layer
+      3. ReLU activation unit
+
+    Args:
+      channels: the number of generated feature.
+      momentum: the momentum of the batch normalization layer.
+    """
+    super(PointNetConv2Layer, self).__init__()
+    self.channels = channels
+    self.momentum = momentum
+
+  def build(self, input_shape):
+    """Builds the layer with a specified input_shape."""
+    self.conv = Conv2D(self.channels, (1, 1), input_shape=input_shape)
+    self.bn = BatchNormalization(momentum=self.momentum)
+
+  # pylint: disable=arguments-differ
+  def call(self, inputs, training):
```

Comment:
Request: can `training` default to `None` like the rest of keras? `BatchNormalization` knows how to handle it :)

Reply:
yes, I meant to fix this, let me take a look, but you have a good point

Reply:
Ok, now I recall. It's funny, because a regular keras.Layer has:
```def call(self, inputs, **kwargs):```
while a BatchNormalization
```def call(self, inputs, training=None):```

I poked a bit deeper, and the second pattern seems a good match

Reply:
yet this is still given linter warnings without the disable...
```
def call(self, inputs, training=None):  # pylint: disable=arguments-differ
```